### PR TITLE
[optin/puller] task: tx sending

### PR DESCRIPTION
**Pull request by gimre-xymcity**

### CODE REVIEW

File: optin/puller/workflows/payout.py [hunk 1]
```diff
@@ -15,10 +20,40 @@
...
+		})
+		size = transaction.size
+		transaction.fee = Amount(size * 1_000)
+
+		signature = self.facade.sign_transaction(self.key_pair, transaction)
+		self.facade.transaction_factory.attach_signature(transaction, signature)
+
+		print(f'tx size {size}')
+		print(transaction)
+
```
Comment (gimre-xymcity) [inline optin/puller/workflows/payout.py:55]: guess should be dropped

File: optin/puller/workflows/payout.py [hunk 2]
```diff
@@ -55,20 +90,38 @@ async def main():
...
+		symbol_network = SymbolNetwork.TESTNET
+
 	client = SymbolClient(args.symbol_node)
 
+	key_pair = get_private_key(Path(args.hot))
+	preparer = TransactionPreparer(args.network, key_pair)
+
+	def send_funds(destination_public_key):
+		destination_address = preparer.facade.network.public_key_to_address(destination_public_key)
+		amount = 0
```
Comment (gimre-xymcity) [inline optin/puller/workflows/payout.py:110]: missing

File: optin/puller/workflows/payout.py [hunk 3]
```diff
@@ -15,10 +20,40 @@
 # TODO_:
 # 2. implement send_funds
```
Comment (Jaguar0625) [inline optin/puller/workflows/payout.py:22]: remove

File: optin/puller/workflows/payout.py [hunk 4]
```diff
@@ -15,10 +20,40 @@
...
 
+XYM_MOSAIC_ID = 0x6BED913FA20223F8
 
-def send_funds(_):
-	# TODO_: implement this
-	return Hash256('')
+
+def get_private_key(filepath):
```
Comment (Jaguar0625) [inline optin/puller/workflows/payout.py:26]: key pair, not private key

File: optin/puller/workflows/payout.py [hunk 5]
```diff
@@ -15,10 +20,40 @@
...
+	password = getpass.getpass(f'Enter password for {filepath}: ')
```
Comment (Jaguar0625) [inline optin/puller/workflows/payout.py:27]: i thought this was supposed to be headless script ?

File: optin/puller/workflows/payout.py (hunk 1 above)
Comment (Jaguar0625) [reply optin/puller/workflows/payout.py:55]: y

File: optin/puller/workflows/payout.py [hunk 6]
```diff
@@ -55,20 +90,38 @@ async def main():
...
+
 	client = SymbolClient(args.symbol_node)
 
+	key_pair = get_private_key(Path(args.hot))
+	preparer = TransactionPreparer(args.network, key_pair)
+
+	def send_funds(destination_public_key):
+		destination_address = preparer.facade.network.public_key_to_address(destination_public_key)
+		amount = 0
+		deadline = client.node_time() + 60 * 60 * 1000
```
Comment (Jaguar0625) [inline optin/puller/workflows/payout.py:111]: maybe should return NetworkTimestamp and then use add_hours(1) ?

File: optin/puller/workflows/payout.py [hunk 7]
```diff
@@ -15,10 +20,40 @@
...
+	storage = PrivateKeyStorage(filepath.parent, password.strip())
+	return KeyPair(storage.load(filepath.stem))
+
+
+class TransactionPreparer:
+	def __init__(self, network_name, key_pair):
+		self.facade = SymbolFacade(network_name)
+		self.key_pair = key_pair
+
+	def prepare_transaction(self, recipient_address, amount, deadline, nem_tx_hash=''):
```
Comment (Jaguar0625) [inline optin/puller/workflows/payout.py:37]: why default '' ... seems it is always provided?

File: optin/puller/workflows/payout.py (hunk 5 above)
Comment (gimre-xymcity) [reply optin/puller/workflows/payout.py:27]: a) I assume we wanted to pass password somehow
b) alternative is via ENV, can't say I'm a fan of that, esp if you have some special chars...

File: optin/puller/workflows/payout.py [hunk 8]
```diff
@@ -15,10 +20,40 @@
...
+			'type': 'transfer_transaction',
+			'signer_public_key': self.key_pair.public_key,
+			'fee': 0,
+			'deadline': deadline,
+			'recipient_address': recipient_address,
+			'message': f'nem tx hash {nem_tx_hash}',
+			'mosaics': [{'mosaic_id': XYM_MOSAIC_ID, 'amount': amount}]
+		})
+		size = transaction.size
+		transaction.fee = Amount(size * 1_000)
```
[the remark this answers is not in the document]
Comment (gimre-xymcity) [reply optin/puller/workflows/payout.py:48]: 100 should be more than enough, but I think we can throw some fees on top(?)

**Comment by Jaguar0625** [comment]
please fix lint

File: optin/puller/workflows/payout.py [hunk 9]
```diff
@@ -55,20 +61,35 @@ async def main():
...
 	nem_network = NemNetwork.MAINNET
 	symbol_network = SymbolNetwork.MAINNET
+	if 'testnet' == args.network:
+		nem_network = NemNetwork.TESTNET
+		symbol_network = SymbolNetwork.TESTNET
+
 	client = SymbolClient(args.symbol_node)
 
+	key_pair = get_key_pair(Path(args.hot))
+	preparer = TransactionPreparer(args.network, client.network_currency(), key_pair)
```
Comment (Jaguar0625) [inline optin/puller/workflows/payout.py:77]: aren't you missing await?

File: optin/puller/tests/processors/test_TransactionPreparer.py [hunk 10]
```diff
@@ -0,0 +1,40 @@
...
+		self.assertEqual(transaction.message, b'nem tx hash 3.141592')
+
+		self.assertEqual(len(transaction.mosaics), 1)
+		self.assertEqual(transaction.mosaics[0].mosaic_id, UnresolvedMosaicId(0x1234_5678_90AB_CDEF))
+		self.assertEqual(transaction.mosaics[0].amount, Amount(1234))
+
+		# - hash
+		expected_hash = preparer.facade.hash_transaction(transaction)
+		self.assertEqual(transaction_hash, Hash256('A342B0065EB6B98D1D96B976A51F2D225DA11D53BC862A86E30089FA70871453'))
+		self.assertEqual(transaction_hash, expected_hash)
```
[the remark this answers is not in the document]
Comment (gimre-xymcity) [reply optin/puller/tests/processors/test_TransactionPreparer.py:40]: I forgot we use that order for assertEquals (somehow thought we do reverse here)

File: optin/puller/puller/processors/TransactionPreparer.py [hunk 11]
```diff
@@ -1,7 +1,15 @@
+import json
+from binascii import hexlify
+
 from symbolchain.facade.SymbolFacade import SymbolFacade
 from symbolchain.sc import Amount
 
 
+def convert_to_dumb_format(nem_address):
```
Comment (gimre-xymcity) [inline optin/puller/puller/processors/TransactionPreparer.py:8]: I assume name makes sense 🙊

File: optin/puller/workflows/payout.py [hunk 12]
```diff
@@ -51,8 +51,7 @@ async def process_unprocessed_requests(unprocessed_requests, sent_requests, data
 			if databases.multisig.check_cosigners(address, [request.address for request in request_group.requests]):
 				balance = databases.balances.lookup_balance(address)
 				symbol_address = symbol_network.public_key_to_address(request_group.requests[0].destination_public_key)
-				nem_transaction_hash = request_group.requests[0].transaction_hash
-				sent_transaction_hash = send_funds(symbol_address, balance, nem_transaction_hash)
+				sent_transaction_hash = send_funds(symbol_address, balance, address)
```
Comment (gimre-xymcity) [inline optin/puller/workflows/payout.py:54]: and here I hope address in grouping is what we want to be in that message, right?

**Comment by Jaguar0625** [comment]
lint is failing, otherwise good

**Review by Jaguar0625 — APPROVED** [review]

**Pull request merged**